File: test_diff_reporter.py
"""
improved test error reporting for the language compiler tests.
handles large diffs gracefully and stores full output for debugging.
"""

import difflib
import tempfile
from pathlib import Path
from typing import Optional, Tuple

class TestDiffReporter:
    """handles test output comparison with better diff reporting."""
    
    def __init__(self, test_case_path: Path):
        self.test_case_path = test_case_path
        self.diff_dir = test_case_path / "test_diffs"
        self.diff_dir.mkdir(exist_ok=True)
    
    def compare_text(self, actual: str, expected: str, output_type: str) -> Tuple[bool, Optional[str]]:
        """
        compare actual vs expected text and return (is_equal, error_message).
        if not equal, stores full diff in test directory and returns helpful error message.
        """
        if actual.strip() == expected.strip():
            return True, None
        
        # store the actual output for debugging
        actual_file = self.diff_dir / f"actual_{output_type}.txt"
        with open(actual_file, 'w', encoding='utf-8') as f:
            f.write(actual)
        
        # store the expected output for comparison
        expected_file = self.diff_dir / f"expected_{output_type}.txt"
        with open(expected_file, 'w', encoding='utf-8') as f:
            f.write(expected)
        
        # generate unified diff
        diff_lines = list(difflib.unified_diff(
            expected.splitlines(keepends=True),
            actual.splitlines(keepends=True),
            fromfile=f"expected_{output_type}",
            tofile=f"actual_{output_type}",
            lineterm=""
        ))
        
        # store full diff
        diff_file = self.diff_dir / f"diff_{output_type}.txt"
        with open(diff_file, 'w', encoding='utf-8') as f:
            f.writelines(diff_lines)
        
        # create error message
        diff_preview = '\n'.join(diff_lines[:50])  # first 50 lines of diff
        if len(diff_lines) > 50:
            diff_preview += f"\n... (and {len(diff_lines) - 50} more lines)\n"
        
        error_msg = f"""{output_type} mismatch.

preview of differences:
{diff_preview}

full diff stored in: {diff_file}
actual output stored in: {actual_file}
expected output stored in: {expected_file}

to review:
  cat {diff_file}
  diff {expected_file} {actual_file}
"""
        
        return False, error_msg

def create_test_diff_reporter(test_case: "TestCase") -> TestDiffReporter:
    """create a diff reporter for the given test case."""
    return TestDiffReporter(test_case.case_path)

File: test_modules/test_diff_reporter.ts
// test_modules/test_diff_reporter.ts

import { TestCase } from "./discovery.ts";

export class TestDiffReporter {
    private testCasePath: string;
    private diffDir: string;

    constructor(testCasePath: string) {
        this.testCasePath = testCasePath;
        this.diffDir = `${testCasePath}/test_diffs`;
        try {
            Deno.mkdirSync(this.diffDir, { recursive: true });
        } catch {
            // ignore
        }
    }

    compareText(actual: string, expected: string, outputType: string): [boolean, string | null] {
        if (actual.trim() === expected.trim()) {
            return [true, null];
        }

        const actualFile = `${this.diffDir}/actual_${outputType}.txt`;
        const expectedFile = `${this.diffDir}/expected_${outputType}.txt`;
        const diffFile = `${this.diffDir}/diff_${outputType}.txt`;

        Deno.writeTextFileSync(actualFile, actual);
        Deno.writeTextFileSync(expectedFile, expected);

        const diffLines = unifiedDiff(
            expected.split(/\r?\n/),
            actual.split(/\r?\n/),
            `expected_${outputType}`,
            `actual_${outputType}`
        );

        Deno.writeTextFileSync(diffFile, diffLines.join("\n"));

        const previewLines = diffLines.slice(0, 50);
        let diffPreview = previewLines.join("\n");
        if (diffLines.length > 50) {
            diffPreview += `\n... (and ${diffLines.length - 50} more lines)\n`;
        }

        const errorMsg =
            `${outputType} mismatch.\n\n` +
            `preview of differences:\n` +
            `${diffPreview}\n\n` +
            `full diff stored in: ${diffFile}\n` +
            `actual output stored in: ${actualFile}\n` +
            `expected output stored in: ${expectedFile}\n\n` +
            `to review:\n` +
            `  cat ${diffFile}\n` +
            `  diff ${expectedFile} ${actualFile}\n`;

        return [false, errorMsg];
    }
}

export function createTestDiffReporter(testCase: TestCase): TestDiffReporter {
    return new TestDiffReporter(testCase.casePath);
}

// very small unified diff implementation
function unifiedDiff(
    a: string[],
    b: string[],
    fromFile: string,
    toFile: string
): string[] {
    const header = [
        `--- ${fromFile}`,
        `+++ ${toFile}`
    ];

    const hunks: string[] = [];

    // naive line-by-line diff, not LCS-perfect but good enough for test hints
    const maxLen = Math.max(a.length, b.length);
    for (let i = 0; i < maxLen; i++) {
        const left = i < a.length ? a[i] : null;
        const right = i < b.length ? b[i] : null;
        if (left === right) {
            hunks.push(` ${left ?? ""}`);
        } else {
            if (left !== null) {
                hunks.push(`-${left}`);
            }
            if (right !== null) {
                hunks.push(`+${right}`);
            }
        }
    }

    return [...header, ...hunks];
}
